File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/asm/AssemblerMultiByteNopTests.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.asm;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import com.github.icedland.iced.x86.Code;
import com.github.icedland.iced.x86.CodeWriterImpl;
import com.github.icedland.iced.x86.Instruction;
import com.github.icedland.iced.x86.dec.ByteArrayCodeReader;
import com.github.icedland.iced.x86.dec.Decoder;

public class AssemblerMultiByteNopTests {
	@Test
	public void testNops() {
		for (int bitness : new int[] { 16, 32, 64 }) {
			for (int i = 0; i <= 128; i++) {
				CodeAssembler a = new CodeAssembler(bitness);
				a.nop(i);
				CodeWriterImpl writer = new CodeWriterImpl();
				a.assemble(writer, 0);
				byte[] data = writer.toArray();
				assertEquals(i, data.length);
				ByteArrayCodeReader reader = new ByteArrayCodeReader(data);
				Decoder decoder = new Decoder(bitness, reader);
				while (reader.canReadByte()) {
					Instruction instr = decoder.decode();
					switch (instr.getCode()) {
					case Code.NOPW:
					case Code.NOPD:
					case Code.NOPQ:
					case Code.NOP_RM16:
					case Code.NOP_RM32:
					case Code.NOP_RM64:
						break;
					default:
						assertTrue(false, String.format("Expected a NOP but got Code %d", instr.getCode()));
						break;
					}
				}
			}

			{
				CodeAssembler a = new CodeAssembler(bitness);
				assertThrows(IllegalArgumentException.class, () -> a.nop(-1));
			}
		}
	}
}

File: src/csharp/Intel/Iced.UnitTests/Intel/AssemblerTests/AssemblerMultiByteNopTests.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if ENCODER && BLOCK_ENCODER && CODE_ASSEMBLER
using System;
using Iced.Intel;
using Xunit;

namespace Iced.UnitTests.Intel.AssemblerTests {
	public class AssemblerMultiByteNopTests {
		[Fact]
		public void TestNops() {
			foreach (var bitness in new[] { 16, 32, 64 }) {
				for (int i = 0; i <= 128; i++) {
					var a = new Assembler(bitness);
					a.nop(i);
					var writer = new CodeWriterImpl();
					a.Assemble(writer, 0);
					var data = writer.ToArray();
					Assert.Equal(i, data.Length);
					var reader = new ByteArrayCodeReader(data);
					var decoder = Decoder.Create(bitness, reader);
					while (reader.CanReadByte) {
						var instr = decoder.Decode();
						switch (instr.Code) {
						case Code.Nopw:
						case Code.Nopd:
						case Code.Nopq:
						case Code.Nop_rm16:
						case Code.Nop_rm32:
						case Code.Nop_rm64:
							break;
						default:
							Assert.Fail($"Expected a NOP but got {instr.Code}");
							break;
						}
					}
				}

				{
					var a = new Assembler(bitness);
					Assert.Throws<ArgumentOutOfRangeException>(() => a.nop(-1));
				}
			}
		}
	}
}
#endif
